tfx/orchestration/experimental/optimizations/beam_fusion.py

```diff
@@ -0,0 +1,197 @@
+# Lint as: python2, python3
+# Copyright 2020 Google LLC. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""TFX BeamFusionOptimizer definition."""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+from __future__ import unicode_literals
+
+from typing import List
+
+from tfx.components.base import base_node
+from tfx.components.base import base_executor
+from tfx.orchestration.pipeline import Pipeline
+
+class BeamFusionOptimizer(object):
+  """Optimizer for TFX pipelines, utilizing Beam Fusion"""
+
+  def __init__(self, pipeline: Pipeline):
+    self.pipeline = pipeline
+
+  def _topologically_sort(self, components: List[base_node.BaseNode],
+                          sources: List[base_node.BaseNode]):
+
+    sorted_components = []
+    current_layer = [c for c in components if c in sources]
+    visited = set(current_layer)
+
+    # Sorts component in topological order.
+    while current_layer:
+      next_layer = []
+      for component in sorted(current_layer, key=lambda c: c.id):
+        sorted_components.append(component)
+        for child in component.downstream_nodes:
+          if child in components and child not in visited:
+            next_layer.append(child)
+            visited.add(child)
+
+      current_layer = next_layer
+
+    return sorted_components
+
+  def _get_intersecting_subgraph(self,
+                                 intersecting_component: base_node.BaseNode,
+                                 fuseable_subgraphs: (
+                                     List[List[base_node.BaseNode]])):
+
+    intersecting_subgraph = None
+    for subgraph in fuseable_subgraphs:
+      if intersecting_component in subgraph:
+        intersecting_subgraph = subgraph
+        break
+
+    return intersecting_subgraph
+
+  def _is_beam_component(self, component: base_node.BaseNode):
+    return issubclass(component.EXECUTOR_SPEC.executor_class,
+                      base_executor.FuseableBeamExecutor)
+
+  def _is_fuseable(self, child: base_node.BaseNode,
+                   current_subgraph: List[base_node.BaseNode]):
+    is_fuseable = True
+
+    # Conduct a BFS to ensure none of the child's ancestors (other than its'
+    # immediate parents and those parents' ancestors) are in current_subgraph.
+    current_layer = []
```

Comment:
I think you can just maintain a queue (or a stack) instead of layers, which would simplify things.

Reply:
Done, used a queue.